### PR TITLE
[gRPC] WalletUnlocker and RPCServer unification

### LINKED ISSUE #3982: Unify WalletUnlocker with main RPC server
Currently lnd will spin up the `WalletUnlocker` gRPC service at startup, allowing the user to unlock the wallet. First when the wallet is unlocked lnd can start up properly, shut down the `WalletUnlocker` and spin up the regular `Lightning` gRPC service, on the same host and port.

It was done this way originally because there was an assumption of the `rpcServer` operating that the wallet was unlocked. This way we could ensure the wallet was in an unlocked state before starting the regular `rpcServer`, not having to change it.

However, this can lead to a few problems that certain users run into:
- Some gRPC clients get confused having two services being run on the same host:port.
- There is no easy way of adding a unified "status API" for getting the status of the wallet, since this would need to be available on both services (see #615).
- It is harder to maintain, as there are two services that need to be tended to, for instance when adding a new credentials option, the logic must be duplicated and added for both services.
- Timing issues, since there is a gap between the wallet unlocker shutting down and the rpc server being ready.

### Steps to completion
- [ ] Merge the two services, such that APIs can all be accessed the moment lnd starts up.
- [ ] Introduce a new `wallet_state` that can be accessed through a new API endpoint that will indicate whether the wallet is `nonexisting`, `locked`, `unlocked`.
- [ ] Start the RPC server immediately when lnd starts, ensuring it can handle the fact that some of its dependencies are not yet created (like the `server`)
- [ ] A way of detecting whether a called API is allowed in the current `wallet_state`. A possible avenue here is to add gRPC interceptor (similar to what we do for macaroons) that checks the current `wallet_state`, and rejects all  `non-WalletUnlocker` calle if the wallet is no `unlocked`.

### PR DESCRIPTION
**Pull request by halseth**
This PR unifies the gRPC server used for the `WalletUnlockerService` and `LightningService`, making both services available at all times. This addresses a long-standing issue of the gRPC server being torn down and restarted after unlocking the wallet.

### Overview
The reason for the original design of tearing down the gRPC server after unlocking the wallet, was that the RPC server (`LightningService`) was dependent on the wallet being unlocked in order to being created. We mitigate this by now creating the RPC server before the wallet is unlocked, in a "non-functioning" state just in order to register its interface with the gRPC server. After the wallet is finally unlocked, we can populate the dependencies of the root RPC server and its subservers, enabling the full `LightningService` to be operational.

### RPC state interception
Since the RPC server shouldn't be used before it has been fully initialized, we gate the calls to the services through a new gRPC interceptor found in the `rpcperms` package. This `RPCInterceptor` keeps track of the current state of the wallet and RPC server, and allows or rejects calls to the various services based on this. We also move macaroon authentication here.

The RPC state tracked by the `RPCInterceptor` also paves the way for a coming new RPC endpoint for exposing the unlocking state of the wallet. 

Fixes #3982 
Fixes https://github.com/lightningnetwork/lnd/issues/4340
Fixes https://github.com/lightningnetwork/lnd/issues/3631

### CODE REVIEW

**Comment by halseth** [comment]
@guggero Can you check if this has some implications for the GRUB setup?

File: rpcperms/interceptor.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+			logger.Errorf("[%v]: %v", info.FullMethod, err)
+		}
+
+		return err
+	}
+}
+
+// macaroonUnaryServerInterceptor is a GRPC interceptor that checks whether the
+// request is authorized by the included macaroons.
+func (r *RPCInterceptor) macaroonUnaryServerInterceptor() grpc.UnaryServerInterceptor {
```
Comment (guggero) [inline rpcperms/interceptor.go:300]: This method (and `macaroonStreamServerInterceptor`) need to be exported because they're currently used by loop, pool and faraday for their macaroon checker.

File: lnd.go [hunk 2]
```diff
@@ -1195,8 +1142,7 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
 	for _, lis := range listeners {
 		wg.Add(1)
 		go func(lis *ListenerWithSignal) {
-			rpcsLog.Infof("Password RPC server listening on %s",
-				lis.Addr())
+			rpcsLog.Infof("RPC server listening on %s", lis.Addr())
 
 			// Close the ready chan to indicate we are listening.
 			close(lis.Ready)
```
Comment (guggero) [inline lnd.go:1163]: Because we now start both listeners at the same time, the `Ready` is closed immediately. We rely on this to only be closed on the main RPC listener. Perhaps we now don't need two separate listeners anymore and can simplify this as well? Would require more refactoring and updates to the falafel template. But it looks like you already started some of this in #5010. We just need two signals (ready to be unlocked and ready for main operation) on the same listener.

**Review by guggero — COMMENTED** [review]
Awesome refactor PR :1st_place_medal: 
This will help many users of our APIs.

I created some test PRs for falafel, loop, pool and LiT and it looks like this should work and not create any issues.
Two things are blocking me to fully test this, see my comments.

Will do a full in-depth review as soon as I was able to confirm this will work with LiT.

**Comment by hsjoberg** [comment]
Agreeing on the awesomeness. I'll have a full review and testing as well.

File: lnd.go (hunk 2 above)
Comment (halseth) [reply lnd.go:1163]: Yeah, the second listener is removed in #5010. 

It is hard to remove it in this PR since it will break the mobile build, and that change relies on a `falafel` change. But if you want to test this , should be able to test #5010 directly, as it build on top of this!

File: rpcperms/interceptor.go (hunk 1 above)
Comment (halseth) [reply rpcperms/interceptor.go:300]: Ah. I guess we would need it to still be defined on the `*Service` also?

Comment (guggero) [reply rpcperms/interceptor.go:300]: You mean the `MacaroonService`? Not necessarily, I can instantiate the `RPCInterceptor` directly and use its methods, so no problem there as long as they're exported.

Comment (halseth) [reply rpcperms/interceptor.go]: Pushed a commit to export them!

File: lnd.go (hunk 2 above)
Comment (guggero) [reply lnd.go:1163]: Ah, I completely missed the new functionality in #5005. Very cool!
I added a PR to `lndclient` and all other projects and was able to verify that this works well in `lightning-terminal`: https://github.com/lightninglabs/lndclient/pull/41
Awesome, this makes it so much easier and cleaner too :tada:

File: lnrpc/autopilotrpc/autopilot_server.go [hunk 3]
```diff
@@ -51,6 +51,13 @@ var (
 	}
 )
 
+// RPCServer is shell struct holding a reference to the actual sub-server. It
```
Comment (guggero) [inline lnrpc/autopilotrpc/autopilot_server.go:54]: nit: is _a_ shell struct.

File: lnrpc/autopilotrpc/autopilot_server.go [hunk 4]
```diff
@@ -51,6 +51,13 @@ var (
...
+// is used to register the gRPC sub-server with the root server before we have
+// the necessary dependencies to populate the actual sub-server.
+type RPCServer struct {
```
Comment (guggero) [inline lnrpc/autopilotrpc/autopilot_server.go:57]: Could we perhaps name this `ServerShell` instead? Would make it a bit easier to distinguish from the actual server.

File: lnrpc/autopilotrpc/autopilot_server.go [hunk 5]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/grpc-ecosystem/grpc-gateway/runtime"
 	"github.com/lightningnetwork/lnd/autopilot"
 	"github.com/lightningnetwork/lnd/lnrpc"
-	"google.golang.org/grpc"
+	grpc "google.golang.org/grpc"
```
Comment (guggero) [inline lnrpc/autopilotrpc/autopilot_server.go:14]: Why is this necessary?

File: rpcperms/interceptor.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	r.svc = svc
+}
+
+// AddPermission adds a new macaroon rule for the given method.
+func (r *RPCInterceptor) AddPermission(method string, ops []bakery.Op) error {
+	r.Lock()
+	defer r.Unlock()
+
+	if _, ok := r.permissionMap[method]; ok {
+		return fmt.Errorf("detected "+
```
Comment (guggero) [inline rpcperms/interceptor.go:140]: nit: can use more of the line now.

File: rpcperms/interceptor.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+	"github.com/lightningnetwork/lnd/macaroons"
+	"github.com/lightningnetwork/lnd/monitoring"
+	"google.golang.org/grpc"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+)
+
+// RPCInterceptor is a struct that can be added to the running GRPC server,
+// intercepting API calls. This is useful for logging, enforcing permissions
+// etc.
+type RPCInterceptor struct {
```
Comment (guggero) [inline rpcperms/interceptor.go:71]: Could we call this `InterceptorChain`? I'm working on something that'll build on top of this and will actually intercept (and potentially manipulate) RPC messages, similar with what we do in the HTLC interceptor (need this for the macaroon accounts feature). So I would also call that `RPCInterceptor`.
I was also trying to come up with a good and short package name. I like `rpcperms` :+1:

File: rpcperms/interceptor.go [hunk 8]
```diff
@@ -156,49 +168,66 @@ func errorLogStreamServerInterceptor(logger btclog.Logger) grpc.StreamServerInte
...
+
+	// Find out if there is an external validator registered for
+	// this method. Fall back to the internal one if there isn't.
+	validator, ok := svc.ExternalValidators[fullMethod]
+	if !ok {
+		validator = svc
+	}
+
+	// Now that we know what validator to use, let it do its work.
+	err := validator.ValidateMacaroon(ctx, uriPermissions, fullMethod)
```
Comment (guggero) [inline rpcperms/interceptor.go:290]: nit: can now return the result directly.

File: lnd.go [hunk 9]
```diff
@@ -1239,16 +1258,37 @@ func waitForWalletPassword(cfg *Config, restEndpoints []net.Addr,
...
 	wg.Wait()
 
+	return shutdown, nil
+}
+
+// waitForWalletPassword blocks until a password is provided by the user to
+// this RPC server.
+func waitForWalletPassword(cfg *Config,
+	pwService *walletunlocker.UnlockerService) (*WalletUnlockParams, error) {
+	chainConfig := cfg.Bitcoin
```
Comment (guggero) [inline lnd.go:1302]: nit: insert newline above.

File: rpcperms/interceptor.go [hunk 10]
```diff
@@ -90,16 +158,16 @@ func (r *RPCInterceptor) Permissions() map[string][]bakery.Op {
 // CreateServerOpts creates the GRPC server options that can be added to a GRPC
 // server in order to add this RPCInterceptor.
 func (r *RPCInterceptor) CreateServerOpts() []grpc.ServerOption {
-	macUnaryInterceptors := []grpc.UnaryServerInterceptor{}
-	macStrmInterceptors := []grpc.StreamServerInterceptor{}
+	var unaryInterceptors []grpc.UnaryServerInterceptor
```
Comment (guggero) [inline rpcperms/interceptor.go:161]: I think this whole section could be simplified even more by just adding all the interceptors to this slice as a static initialization.

File: rpcperms/interceptor.go [hunk 11]
```diff
@@ -120,6 +188,14 @@ func (r *RPCInterceptor) CreateServerOpts() []grpc.ServerOption {
...
+	unaryInterceptors = append(
+		unaryInterceptors, r.rpcStateUnaryServerInterceptor(),
+	)
+	strmInterceptors = append(
+		strmInterceptors, r.rpcStateStreamServerInterceptor(),
+	)
+
 	// If any interceptors have been set up, add them to the server options.
 	var serverOpts []grpc.ServerOption
 	if len(unaryInterceptors) != 0 && len(strmInterceptors) != 0 {
```
Comment (guggero) [inline rpcperms/interceptor.go:201]: We know none of them are empty as we've just added a bunch above.

File: rpcperms/interceptor.go [hunk 12]
```diff
@@ -120,6 +188,14 @@ func (r *RPCInterceptor) CreateServerOpts() []grpc.ServerOption {
 		strmInterceptors, errorLogStreamServerInterceptor(r.rpcsLog),
 	)
 
+	// Add RPC state check interceptors.
+	unaryInterceptors = append(
+		unaryInterceptors, r.rpcStateUnaryServerInterceptor(),
```
Comment (guggero) [inline rpcperms/interceptor.go:193]: Should the state interceptors come before the macaroon interceptors?

File: lnd.go [hunk 13]
```diff
@@ -334,18 +334,26 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
 	rpcServerOpts := rpcInterceptor.CreateServerOpts()
 	serverOpts = append(serverOpts, rpcServerOpts...)
 
 	grpcServer := grpc.NewServer(serverOpts...)
 	defer grpcServer.Stop()
 
-	// We'll create the WalletUnlockerService and register this with the
-	// GRPC server.
-	pwService := createWalletUnlockerService(cfg)
+	// Register the  WalletUnlockerService with the GRPC server.
```
Comment (guggero) [inline lnd.go:356]: nit: extra space.

**Review by guggero — COMMENTED** [review]
Great refactor and cleanup PR :1st_place_medal: 
This will make many things much easier, thank you for tackling this.

I don't have any major comments, it's mostly about naming and some small simplifications. I was also able to confirm this works great for LiT (in combination with the follow-up PRs).

File: lnrpc/autopilotrpc/autopilot_server.go (hunk 4 above)
Comment (halseth) [reply lnrpc/autopilotrpc/autopilot_server.go]: Good idea, done!

File: lnrpc/autopilotrpc/autopilot_server.go (hunk 5 above)
Comment (halseth) [reply lnrpc/autopilotrpc/autopilot_server.go]: Not sure, this was done automatically by the import tool. Reverted.

File: rpcperms/interceptor.go (hunk 7 above)
Comment (halseth) [reply rpcperms/interceptor.go]: I like it, done!

File: rpcperms/interceptor.go (hunk 10 above)
Comment (halseth) [reply rpcperms/interceptor.go:159]: I actually like it this way, easier to see that we add them one by one?

File: rpcperms/interceptor.go (hunk 11 above)
Comment (halseth) [reply rpcperms/interceptor.go]: Haha, true! Did just copy this, but you're right this if is not needed anymore 👍

File: rpcperms/interceptor.go (hunk 12 above)
Comment (halseth) [reply rpcperms/interceptor.go]: I think you are right, that makes more sense! Done.

I also moved the logging interceptors first, makes sense that all calls are logged, whether they pass authentication or not.

**Comment by guggero** [comment]
Fails because of a bunch of errors that need to be white listed:

```
<time> [ERR] RPCS: [/lnrpc.Lightning/BakeMacaroon]: permission denied
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: cannot retrieve macaroon: cannot get macaroon: root key with id  doesn't exist
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: caveat "ipaddr 1.1.1.1" not satisfied: macaroon locked to different IP address
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: caveat "time-before 2021-02-25T<chan>.601693549Z" not satisfied: macaroon has expired
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: expected 1 macaroon, got 0
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: permission denied
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: RPC starting up
<time> [ERR] RPCS: [/lnrpc.Lightning/ListMacaroonIDs]: cannot retrieve macaroon: cannot get macaroon: root key with id 1 doesn't exist
<time> [ERR] RPCS: [/lnrpc.Lightning/NewAddress]: permission denied
```

File: rpcperms/interceptor.go [hunk 14]
```diff
@@ -151,49 +163,61 @@ func errorLogStreamServerInterceptor(logger btclog.Logger) grpc.StreamServerInte
 	}
 }
 
+// checkMacaroon validates that the context contains the macaoon needed to
```
Comment (guggero) [inline rpcperms/interceptor.go:245]: nit: s/macaoon/macaroon/

**Review by guggero — APPROVED** [review]
LGTM, excited for this to land :rocket: 

Will do more manual tests of edge cases in combination with the follow-up PRs.

File: rpcperms/interceptor.go [hunk 15]
```diff
@@ -7,13 +7,53 @@ import (
...
+	// RPC server is not yeat ready.
+	walletUnlocked
+
+	// rpcActive means that the RPC server is ready to accept calls.
+	rpcActive
+)
+
 var (
+	// ErrNoWallet is returned if the wallet does not exist.
+	ErrNoWallet = fmt.Errorf("wallet not created")
```
Comment (wpaulino) [inline rpcperms/interceptor.go:43]: Nit: Some error message suggestions:
* "wallet not created, create one to enable full RPC access"
* "wallet locked,  unlock it to enable full RPC access"
* "wallet already unlocked, WalletUnlocker service is no longer available"

File: rpcperms/interceptor.go [hunk 16]
```diff
@@ -288,9 +288,9 @@ func (r *InterceptorChain) checkMacaroon(ctx context.Context,
 	return validator.ValidateMacaroon(ctx, uriPermissions, fullMethod)
 }
 
-// macaroonUnaryServerInterceptor is a GRPC interceptor that checks whether the
+// MacaroonUnaryServerInterceptor is a GRPC interceptor that checks whether the
```
Comment (wpaulino) [inline rpcperms/interceptor.go:291]: Nit: Include rationale for exporting in commit message.

**Review by wpaulino — COMMENTED** [review]
Great work to address a common pain point of many users! Diff reads well to me, did some manual testing and didn't run into any issues.

File: rpcperms/interceptor.go [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+	r.permissionMap[method] = ops
+	return nil
+}
+
+// Permissions returns the current set of macaroon permissions.
+func (r *InterceptorChain) Permissions() map[string][]bakery.Op {
+	r.RLock()
+	defer r.RUnlock()
+
+	return r.permissionMap
```
[the remark this answers is not in the document]
Comment (halseth) [reply rpcperms/interceptor.go]: Yep, good catch!

**Review by wpaulino — APPROVED** [review]
Looks like there's one more error that needs to be whitelisted.

```
<time> [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: caveat "time-before 2021-03-10T<chan>.07603761Z" not satisfied: macaroon has expired
```

**Comment by halseth** [comment]
Thanks for the review! 

It's almost a bit sentimental for me to hit the merge button on this one, it's been long time in the making, and required many attempts before we settled on a non-breaking change that also wasn't a pain to review 😭 

❤️

**Comment by Kixunil** [comment]
Thanks a lot! Sorry for not testing it, unfortunately, my life took weird turns lately.
